Minor performance and style improvements

eduherminio opened the pull request:
Feel free to cherry-pick only some of the commits (or none of them!)

All these changes have been detected by static analysis and auto-resolved using VS suggestions, otherwise I wouldn't date to send such a length PR.

* `Count()` -> `.Count()` or `.Length` when possible
* `Any()` -> `.Count()` or `.Length` when possible
* Remove setters when possible
* Use pattern matching when possible (avoiding double casting) 
* Simplify string comparison (`x.ToLowerInvariant() == y` -> `string.Equals(x, y,  StringComparison.InvariantCultureIgnoreCase)`)
* Simplify collection initialization when possible
* Protected constructor in abstract class
* Remove unnecessarily interpolated strings
* Simplify always true exceptions (https://github.com/OpenRA/OpenRA/commit/b805481aac0b95485c397290e761d884abbed5e1, please review carefully)

eduherminio committed:
.Any(), .Count()  -> .Count or .Length

eduherminio committed:
Use read-only autoimplemented property when possible

eduherminio committed:
Remove unnecessarily interpolated strings

eduherminio committed:
Use pattern matching

eduherminio committed:
Convert extension in real extension

eduherminio committed:
Dictionary optimization

eduherminio committed:
Abstract class with public constructor

eduherminio committed:
Optimize string comparison

eduherminio committed:
Simplify collection initialization

eduherminio committed:
Simplify 'always true' expressions

Co-authored-by: atlimit8 <atlimit8-vcs@gmx.com>

eduherminio committed:
Include name in authors

abcdefg30 commented on OpenRA.Game/Primitives/BitSet.cs:
```diff
@@ -98,7 +98,7 @@ public override string ToString()
 
 		public bool Equals(BitSet<T> other) { return other == this; }
-		public override bool Equals(object obj) { return obj is BitSet<T> && Equals((BitSet<T>)obj); }
+		public override bool Equals(object obj) { return obj is BitSet<T> bitSets && Equals(bitSets); }
```
```suggestion
		public override bool Equals(object obj) { return obj is BitSet<T> bitSet && Equals(bitSet); }
```
?
(This is only a single `BitSet` isn't it.)

abcdefg30 commented on OpenRA.Mods.Cnc/Activities/Teleport.cs:
```diff
@@ -133,7 +133,7 @@ public override bool Tick(Actor self)
 			{
 				if (teleporter.Owner.Shroud.IsExplored(tile)
-					&& (restrictTo == null || (restrictTo != null && restrictTo.Contains(tile)))
+					&& (restrictTo == null || (restrictTo.Contains(tile)))
```
```suggestion
					&& (restrictTo == null || restrictTo.Contains(tile))
```

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Minelayer.cs:
```diff
@@ -146,11 +146,10 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				var minefield = GetMinefieldCells(minefieldStart, cell, Info.MinefieldDepth)
 					.Where(c => IsCellAcceptable(self, c) && self.Owner.Shroud.IsExplored(c)
-						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (mobile != null && mobile.CanStayInCell(c)))
+						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (movement is Mobile mobile && mobile.CanStayInCell(c)))
```
I'm pretty sure the cast was done outside of the Linq clause to avoid having to do it every time the `Where` runs.

eduherminio replied:
You're right, I've reverted the change

abcdefg30 commented on OpenRA.Game/Graphics/ModelRenderer.cs:
```diff
@@ -79,7 +79,7 @@ public void SetViewportParams()
 		}
 
-		public ModelRenderProxy RenderAsync(
+		public ModelRenderProxy Render(
```
This isn't using any `async` code in the literal sense, but doesn't do any rendering right now either if I understand correctly (since it is appending to `doRender`). So it might make sense to keep this. 🤔

eduherminio replied:
True, reverted.

abcdefg30 commented on OpenRA.Mods.Common/UtilityCommands/ImportLegacyMapCommand.cs:
```diff
@@ -393,7 +390,7 @@ public void LoadActors(IniFile file, string section, List<string> players, Map m
 				{
 					var parts = s.Value.Split(',');
-					if (parts[0] == "")
+					if (parts[0]?.Length == 0)
```
```suggestion
					if (string.IsNullOrEmpty(parts[0]))
```

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyOptionsLogic.cs:
```diff
@@ -83,14 +83,14 @@ void RebuildOptions()
 					foreach (var child in row.Children)
-						if (child is CheckboxWidget)
-							checkboxColumns.Enqueue((CheckboxWidget)child);
+						if (child is CheckboxWidget childe)
+							checkboxColumns.Enqueue(childe);
```
```suggestion
						if (child is CheckboxWidget checkbox)
							checkboxColumns.Enqueue(checkbox);
```

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyOptionsLogic.cs:
```diff
@@ -112,14 +112,14 @@ void RebuildOptions()
 					foreach (var child in row.Children)
-						if (child is DropDownButtonWidget)
-							dropdownColumns.Enqueue((DropDownButtonWidget)child);
+						if (child is DropDownButtonWidget childe)
+							dropdownColumns.Enqueue(childe);
```
```suggestion
						if (child is DropDownButtonWidget dropDown)
							dropdownColumns.Enqueue(dropDown);
```

abcdefg30 reviewed (commented):
Changes/Ideas look good to me overall, just a few comments. We'll also want to squash the commits to a lower number in the end.

eduherminio commented:
> Changes/Ideas look good to me overall, just a few comments. We'll also want to squash the commits to a lower number in the end.

I (think I) addressed all the comments, do you want me to try to logically groups commits or just squash everything when you merge it?

RoosterDragon commented:
For any issues that are fully resolved, can we add the corresponding `dotnet_diagnostic.XXYYYY.severity = warning` line into `.editorconfig` so they get flagged automatically in the future?

abcdefg30 commented:
> do you want me to try to logically groups commits or just squash everything when you merge it?

Ideal would be one commit per "change", i.e. one for `.Any() -> .Count or .Length`, one for removing the private setters etc. (But if that is too much work for you, I also wouldn't mind if you squashed it down to fewer commits.)

eduherminio committed:
Add fixed issues as warnings to .editorconfig

eduherminio commented:
> > do you want me to try to logically groups commits or just squash everything when you merge it?
> 
> Ideal would be one commit per "change", i.e. one for `.Any() -> .Count or .Length`, one for removing the private setters etc. (But if that is too much work for you, I also wouldn't mind if you squashed it down to fewer commits.)

Done @abcdefg30, also added @RoosterDragon's suggestion

abcdefg30 commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -18,413 +18,411 @@
+        public readonly Size TileSize;
+        public readonly int TileScale;
+        public readonly World World;
+        public Viewport Viewport { get; }
```
Style nit: Spaces -> Tabs

eduherminio replied:
My bad, fixed

Mailaender commented on OpenRA.Game/FieldLoader.cs:
```diff
@@ -564,7 +564,7 @@ public static void Load(object self, MiniYaml my)
 			}
 
-			if (missing.Any())
+			if (missing.Count > 0)
```
We used `Any()` deliberately in the past for performance reasons to not iterate over the whole enumerable.

eduherminio replied:
`Any()` vs `Count()` or vs `Count`?
Using `Count` (property) is [orders of magnitude](https://www.tabsoverspaces.com/233649-comparing-speed-of-count-0-and-any) faster than using any of the LINQ methods and doesn't iterate over the whole enumerable.
You can argue about readability though.

Mailaender replied:
Ah, so `Count` is static. Then let's use that if LINQ is slower.

abcdefg30 commented:
Can you squash the space -> tab fixup into the "Use pattern matching" commit (where I think the change first happens)? (So that we don't have random revert commits in the history.) LGTM afterwards.

eduherminio commented:
> Can you squash the space -> tab fixup into the "Use pattern matching" commit (where I think the change first happens)? (So that we don't have random revert commits in the history.) LGTM afterwards.

👍🏼 Squashed and rebased.

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Minelayer.cs:
```diff
@@ -150,7 +150,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				var minefield = GetMinefieldCells(minefieldStart, cell, Info.MinefieldDepth)
 					.Where(c => IsCellAcceptable(self, c) && self.Owner.Shroud.IsExplored(c)
-						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (mobile != null && mobile.CanStayInCell(c)))
+						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && mobile?.CanStayInCell(c) == true)
```
```suggestion
						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && mobile?.CanStayInCell(c))
```
(Commit "Use pattern matching")

eduherminio replied:
I'm afraid we need that

![image](https://user-images.githubusercontent.com/11148519/167621337-bb79156a-6a03-49d3-afe9-28b761f5d099.png)

pchote replied:
IMO `(mobile != null && mobile.CanStayInCell(c))` is clearer than `mobile?.CanStayInCell(c) == true` due to this exact confusion.

abcdefg30 replied:
I too prefer the old variant over having `== true`. Can you change it back please?

pchote replied:
A compromise could be `(mobile?.CanStayInCell(c) ?? false)` if reverting this change means having to relax one of the new style rules.

eduherminio replied:
I finally went for the original solution

abcdefg30 commented on OpenRA.Game/Network/Handshake.cs:
```diff
@@ -77,10 +77,12 @@ public static HandshakeResponse Deserialize(string data)
-			data.Add(new MiniYamlNode("Client", FieldSaver.Save(Client)));
+			var data = new List<MiniYamlNode>
+			{
+				new MiniYamlNode("Handshake", null,
+				new[] { "Mod", "Version", "Password", "Fingerprint", "AuthSignature", "OrdersProtocol" }.Select(p => FieldSaver.SaveField(this, p)).ToList()),
```
```suggestion
				new MiniYamlNode("Handshake", null,
					new[] { "Mod", "Version", "Password", "Fingerprint", "AuthSignature", "OrdersProtocol" }.Select(p => FieldSaver.SaveField(this, p)).ToList()),
```
(Commit " Simplify collection initialization")

abcdefg30 reviewed (commented):
Sorry, found two more small things.

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Minelayer.cs:
```diff
@@ -328,7 +328,7 @@ protected override IEnumerable<IRenderable> RenderAboveShroud(WorldRenderer wr,
 					}
 					else if (!this.minelayer.IsCellAcceptable(minelayer, c)
-						|| !movement.CanEnterCell(c, null, BlockedByActor.Immovable) || (mobile != null && !mobile.CanStayInCell(c)))
+						|| !movement.CanEnterCell(c, null, BlockedByActor.Immovable) || (mobile?.CanStayInCell(c) == false))
```
Hm, this should probably be `?? true` then to be consistent with above?

eduherminio replied:
Rather `!mobile?.CanStayInCell(c) ?? false`, right?
I think this is far less clear than the original or the existing change, though

abcdefg30 replied:
`?? false`, right. If we don't want to do that I prefer the original version.

eduherminio replied:
I've reverted both changes and suppressed the warning in the source file.

abcdefg30 replied:
We had a short discussion about this on Discord and would prefer to just drop the style warning instead of locally suppressing it.

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Minelayer.cs:
```diff
@@ -150,7 +150,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				var minefield = GetMinefieldCells(minefieldStart, cell, Info.MinefieldDepth)
 					.Where(c => IsCellAcceptable(self, c) && self.Owner.Shroud.IsExplored(c)
-						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (mobile != null && mobile.CanStayInCell(c)))
+						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (mobile?.CanStayInCell(c) ?? false)
```
There is a missing brace now:
```suggestion
						&& movement.CanEnterCell(c, null, BlockedByActor.Immovable) && (mobile?.CanStayInCell(c) ?? false))
```

atlimit8 commented on OpenRA.Mods.Cnc/FileFormats/BlowfishKeyProvider.cs:
```diff
@@ -90,9 +90,8 @@ static void KeyToBigNum(uint[] n, byte[] key, uint len)
-			uint i;
-			i = len - 1;
-			while ((i >= 0) && (n[i] == 0)) i--;
+			uint i = len - 1;
```
##### Minor nit:
`var` should be used since `len` is already a `uint`, "Use `var` when declaring temporary variables inside methods." is part of the [coding standard](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard), and the code is being refactored already. Otherwise, this code may never be refactored towards the coding standard.
```suggestion
			var i = len - 1;
```

eduherminio replied:
I understand that's the nit @atlimit8, any other issue?

atlimit8 reviewed (changes requested):
I found one probable issue and a minor nit.

atlimit8 commented on OpenRA.Mods.Cnc/FileFormats/BlowfishKeyProvider.cs:
```diff
@@ -90,9 +90,8 @@ static void KeyToBigNum(uint[] n, byte[] key, uint len)
-			i = len - 1;
-			while ((i >= 0) && (n[i] == 0)) i--;
+			var i = len - 1;
+			while (n[i] == 0) i--;
```
`while (n[i] == 0) i--;` throws an out of range exception if `n` is zeros `n[len]` through `n[0]`. If that is guaranteed to never happen, then I agree that `while (n[i] == 0) i--;` would be faster.
```suggestion
			while ((i >= 0) && (n[i] == 0)) i--;
```

eduherminio replied:
>  if `n` is zeros `n[len]` through `n[0]`

I'm not sure I understand what you mean there, but it's guaranteed that `i >= 0` because i is an `uint`.
Plus it's detected automatically as a  `'always true' expression`, otherwise I wouldn't dare to touch it.

That's why after the refactoring I decided to keep the explicit `uint` declaration (your nit above), since it helps understanding why `i` can never be `< 0` by design.

atlimit8 replied:
Decrementing a `uint` when it is zero causes it to wrap around to the max value. I have this test code right here:
```c#
var someArray = new uint[] { 0, 0, 8, 1, 0, 0 };
var i = 2U - 1;
while (someArray[i] == 0) i--;
```

atlimit8 replied:
This is actually a bug here! Try:
```c#
var someArray = new uint[] { 0, 0, 8, 1, 0, 0 };
var i = 2U - 1;
while (someArray[i] == 0) i--;
```
I'm removing the fixup request since fixing this bug is out of scope.

atlimit8 reviewed (changes requested):
NOTE: I don't know why my comment for this line was not added to my last review; I probably forgot to add it to the review. That is why I have been waiting.

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/df86d933938b7f9cf340183026e2f7b55b867246)
